<a href="https://colab.research.google.com/github/rameez16/Deep-Learning-Projects/blob/main/Breast_Cancer_Prediction_With_ANN(Classification).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



## 📌 Project Title:
**Breast Cancer Classification using Artificial Neural Network**

---

## 🎯 Objective:
To build and evaluate an Artificial Neural Network (ANN) that can accurately predict whether a tumor is **malignant** or **benign** based on various numerical features extracted from breast mass cell images.

---

## 📊 Dataset Overview:
- **Source**: Breast Cancer Wisconsin (Diagnostic) Dataset
- **Entries**: 569 samples
- **Features**: 30 numerical features (e.g., mean radius, texture, area, smoothness)
- **Target**: Tumor diagnosis label  
  - `Malignant = 1`  
  - `Benign = 0`

---

## 🔍 Features Used:
Statistics including:
- **Mean values**: `mean radius`, `mean texture`, etc.
- **Standard errors**: `radius error`, `texture error`, etc.
- **Worst-case values**: `worst perimeter`, `worst symmetry`, etc.

---

## 🧪 Steps Involved:

### 1. Data Preprocessing
- Load and explore the dataset
- Normalize/scale features
- Encode labels if necessary
- Train-test split (e.g., 80-20)

### 2. Model Building
- Define a Sequential ANN using TensorFlow/Keras
- Input layer: 30 neurons (one per feature)
- Hidden layers: 1–2 layers with **ReLU** activation
- Output layer: 1 neuron with **sigmoid** activation (binary classification)

### 3. Model Compilation
- **Loss function**: `binary_crossentropy`
- **Optimizer**: `adam`
- **Metrics**: `accuracy`

### 4. Training
- Fit model with training data
- Validate with test data
- Use callbacks like `EarlyStopping` (optional)

### 5. Evaluation
- Accuracy
- Confusion matrix
- Precision, Recall, F1-score
- ROC Curve and AUC

### 6. Conclusion
- Summarize model performance
- Suggest improvements and future work

---

## 🧰 Tools & Libraries:
- **Python**
- **Pandas**, **NumPy**
- **Scikit-learn** (for preprocessing and evaluation)
- **TensorFlow / Keras** (for ANN model)
- **Matplotlib**, **Seaborn** (for visualization)

---

## ✅ Expected Outcome:
- A trained ANN model that classifies breast tumors as **malignant** or **benign** with high accuracy (typically >90%)
- Insights into which features contribute most to the diagnosis


DOMAIN KNOWLEDGE



## 🔬 Benign Tumors
- **Definition**: Non-cancerous growths.
- **Growth Rate**: Usually slow.
- **Spread**: Do **not** spread to other parts of the body (non-invasive).
- **Appearance**: Cells look relatively normal under a microscope.
- **Danger Level**: Often not life-threatening unless they press on vital organs.
- **Treatment**: Can often be removed surgically and rarely come back.

**Examples**: Fibroids in the uterus, lipomas (fatty lumps), adenomas.

---

## 🧬 Malignant Tumors (Cancer)
- **Definition**: Cancerous growths that can invade and destroy surrounding tissue.
- **Growth Rate**: Often grow quickly.
- **Spread**: Can spread (metastasize) to distant parts of the body via blood or lymph.
- **Appearance**: Cells are irregular and look abnormal.
- **Danger Level**: Potentially life-threatening if not treated.
- **Treatment**: May require surgery, chemotherapy, radiation, or targeted therapy.

**Examples**: Breast cancer, lung cancer, leukemia, lymphoma.

---

## 🔍 Key Differences Summary

| Feature              | Benign Tumor           | Malignant Tumor         |
|----------------------|------------------------|--------------------------|
| **Growth**           | Slow                   | Rapid                    |
| **Spread**           | No                     | Yes (metastasis)         |
| **Cell appearance**  | Normal-like            | Abnormal and varied      |
| **Treatment**        | Usually surgery        | Surgery + chemo/radiation|
| **Recurrence**       | Rare                   | Possible/likely          |
| **Life-threatening** | Rarely                 | Often                    |

# 1. Data Preprocessing

In [41]:
import numpy as np
import pandas as pd
import sklearn.datasets
from sklearn.model_selection import train_test_split

In [2]:
breast_cancer = sklearn.datasets.load_breast_cancer()


In [3]:
print(breast_cancer)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [4]:
df=pd.DataFrame(breast_cancer.data,columns=breast_cancer.feature_names)

In [5]:
df.head(5)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [6]:
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [8]:
df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
dtype: int64

In [9]:
df['label']=breast_cancer.target

In [10]:
df['label'].value_counts()

label
1    357
0    212
Name: count, dtype: int64

**Data is Imbalenced**

* 1-->Benign Cancre

* 0-->Malignant Cancre

In [11]:
df.shape

(569, 31)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

### **Seprating the features and target**

In [13]:
x=df.drop(columns='label',axis=1)
y=df['label']

### Splitting the data

In [14]:
X_train,X_test,Y_train,Y_test=train_test_split(x,y,test_size=0.2,random_state=2)

In [15]:
X_train.shape

(455, 30)

In [16]:
Y_train.shape

(455,)

### Standardise the Data

In [17]:
from sklearn.preprocessing import StandardScaler

scalar=StandardScaler()

X_train=scalar.fit_transform(X_train)
X_test=scalar.transform(X_test)

#2. Model Building ( Nural NetWork )


 Define a Sequential ANN using TensorFlow/Keras
- Input layer: 30 neurons (one per feature)
- Hidden layers: 1–2 layers with **ReLU** activation
- Output layer: 1 neuron with **sigmoid** activation (binary classification)

In [18]:
import tensorflow as tf
from tensorflow import keras

In [31]:
model=keras.Sequential([keras.layers.Flatten(input_shape=(30,)),
                        keras.layers.Dense(20,activation='relu'),
                        keras.layers.Dense(30,activation='relu'),
                        keras.layers.Dense(20,activation='relu'),
                        keras.layers.Dense(1,activation='sigmoid')])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [32]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_1 (Flatten)             │ (None, 30)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 20)             │           620 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 30)             │           630 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 20)             │           620 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 1)              │            21 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,891 (7.39 KB)

 Trainable params: 1,891 (7.39 KB)

 Non-trainable params: 0 (0.00 B)

### 3. Model Compilation
- **Loss function**: `binary_crossentropy`
- **Optimizer**: `adam`
- **Metrics**: `accuracy`

In [33]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

## 4. Training and Prediction
- Fit model with training data
- Validate with test data

In [34]:
model.fit(X_train,Y_train,validation_split=.1,epochs=10)

Epoch 1/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.3092 - loss: 0.7650 - val_accuracy: 0.8696 - val_loss: 0.5816
Epoch 2/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8001 - loss: 0.5630 - val_accuracy: 0.9348 - val_loss: 0.4363
Epoch 3/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9286 - loss: 0.4364 - val_accuracy: 0.9565 - val_loss: 0.3239
Epoch 4/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9331 - loss: 0.3151 - val_accuracy: 0.9565 - val_loss: 0.2374
Epoch 5/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.9716 - loss: 0.2267 - val_accuracy: 0.9348 - val_loss: 0.1774
Epoch 6/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9687 - loss: 0.1753 - val_accuracy: 0.9348 - val_loss: 0.1380
Epoch 7/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9816 - loss: 0.1311 - val_accuracy: 0.9348 - val_loss: 0.1156
Epoch 8/10
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9943 - loss: 0.0968 - val_accuracy: 0.9348 - val_los

Prediction on Test Data

In [49]:
test_predict=model.predict(X_test)

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 


In [50]:
test_predict[0]

array([0.8268795], dtype=float32)

In [51]:
test_predict = (test_predict >= 0.5).astype(int)

In [52]:
test_predict[0]

array([1])

Prediction on Train data

In [45]:
train_predict=model.predict(X_train)

15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 


In [53]:
train_predict = (train_predict >= 0.5).astype(int)

In [54]:
train_predict[0]

array([1])

# 4. Evaluation

In [55]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, f1_score, precision_score, recall_score

In [56]:
print('Training model accuracy :- ', accuracy_score(Y_train, train_predict))
print('training model Confusion Matrix :- \n',confusion_matrix(Y_train, train_predict))
print('training model precision_score Report :- ',precision_score(Y_train, train_predict))
print('training model recall_score Report :- ',recall_score(Y_train, train_predict))
print(' ')
print('*'*80)
print(' ')
print('Testing model accuracy :- ', accuracy_score(Y_test, test_predict))
print('Testing model Confusion Matrix :- \n',confusion_matrix(Y_test, test_predict))
print('Testing model precision_score Report :- ',precision_score(Y_test, test_predict))
print('Testing model recall_score Report :- ',recall_score(Y_test, test_predict))

Training model accuracy :-  0.9824175824175824
training model Confusion Matrix :- 
 [[162   5]
 [  3 285]]
training model precision_score Report :-  0.9827586206896551
training model recall_score Report :-  0.9895833333333334
 
********************************************************************************
 
Testing model accuracy :-  0.9473684210526315
Testing model Confusion Matrix :- 
 [[41  4]
 [ 2 67]]
Testing model precision_score Report :-  0.9436619718309859
Testing model recall_score Report :-  0.9710144927536232


##  Training Performance

- **Accuracy**: 98.24%
- **Confusion Matrix**:
[[162 5]
[ 3 285]]

markdown
Copy
Edit
- **True Negatives (TN)**: 162
- **False Positives (FP)**: 5
- **False Negatives (FN)**: 3
- **True Positives (TP)**: 285

- **Precision**: 98.28%  
*(High precision indicates very few false positives)*
- **Recall**: 98.96%  
*(High recall means most malignant cases were correctly identified)*

---

##  Testing Performance

- **Accuracy**: 94.74%
- **Confusion Matrix**:
[[41 4]
[ 2 67]]

- **True Negatives (TN)**: 41
- **False Positives (FP)**: 4
- **False Negatives (FN)**: 2
- **True Positives (TP)**: 67

- **Precision**: 94.37%  
- **Recall**: 97.10%

---

## 📈 Summary

| Metric       | Training Set | Testing Set |
|--------------|--------------|-------------|
| **Accuracy** | 98.24%       | 94.74%      |
| **Precision**| 98.28%       | 94.37%      |
| **Recall**   | 98.96%       | 97.10%      |

---

## 🔍 Conclusion

- The model demonstrates excellent performance on the training set, indicating effective learning.
- The slight drop in performance on the test set is expected and indicates generalization ability without overfitting.
- Both precision and recall are high, suggesting that the model is effective at minimizing false positives and false negatives — a critical aspect in medical diagnosis.

---
